### PR TITLE
chore: create Gitpod configuration

### PR DESCRIPTION
**Pull request by HonkingGoose**
## Changes:

- Use a simple Gitpod configuration
  - Use `nvm` to "pin" node to a specific major version
- Document how to use Gitpod (or Codespaces) in `docs/development/remote-development.md`

## Context:

This works okay-ish, but it can probably be improved a lot by somebody competent with Docker.

The `.gitpod.Dockerfile` is cobbled together from the Gitpod programming language reference examples:

- Java [^gitpod-docs-java]
- JavaScript [^gitpod-docs-javascript]
- Python [^gitpod-docs-python]

The Gitpod docs are licensed under the MIT license, so not sure if that is compatible with our license or not. [^gitpod-docs-license]

Docker recommends that you use "multi-stage" builds [^docker-best-practice-multistage] to reduce container size, but I have no clue how to do that properly, so I'm staying away from that myself.

## TODO:

- [x] Improve Gitpod configuration based on review feedback
- [ ] Figure out why `yarn jest` has 2 failing tests when running on Gitpod, but no failing tests on GitHub Actions
- [x] Document how to use Gitpod in our Renovate developer documentation

## Documentation (please check one with an [x])

- [x] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [x] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

[^gitpod-docs-java]: https://www.gitpod.io/docs/languages/java#sdkman-in-docker
[^gitpod-docs-javascript]: https://www.gitpod.io/docs/languages/javascript#node-versions
[^gitpod-docs-python]: https://www.gitpod.io/docs/languages/python#python-versions
[^docker-best-practice-multistage]: https://docs.docker.com/develop/develop-images/dockerfile_best-practices/#use-multi-stage-builds
[^gitpod-docs-license]: https://github.com/gitpod-io/website/blob/main/LICENSE

### CODE REVIEW

**Review by HonkingGoose — COMMENTED** [review]
Questions about `RUN` command ordering and contents.

File: .gitpod.yml [hunk 1]
```diff
@@ -0,0 +1,14 @@
+image:
```
[the remark this answers is not in the document]
Comment (viceice) [reply .gitpod.yml:1]: I think gitpod will set the git commit and author env vars, which override any config file. I've need to override it in my jenkins drone ci pipelines for real runs too.

File: .gitpod.Dockerfile [hunk 2]
```diff
@@ -0,0 +1,21 @@
+FROM gitpod/workspace-full
+
+# Install latest version of Git
+
+RUN sudo apt-get update && sudo apt-get install -y \
```
[the remark this answers is not in the document]
Comment (viceice) [reply .gitpod.Dockerfile:5]: Is the base image based on ubuntu? So we can install out buildpack files and use those installers to install our required versions?

File: .dockerignore [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+**/gradle-wrapper/__fixtures__/**/.gradle
+.clinic/
+
+# generated code
+**/*.generated.ts
+tools/dist
+test/datasource/nuget/_fixtures/obj
+
+# testing
+/docker-compose.yml
```
Comment (viceice) [inline .dockerignore:27]:
```suggestion
*
```

We don't copy any files, so ignore all (Dockerfile is excluded automatically)

File: .gitpod.Dockerfile [hunk 4]
```diff
@@ -0,0 +1,21 @@
+FROM gitpod/workspace-full
```
[the remark this answers is not in the document]
Comment (viceice) [reply .gitpod.Dockerfile:1]: Can we move this to a `.gitpod` folder? so we don't pollute repo root to much?

@rarkins What about a new builder repo to prebuild the devcontainer and gitpod image and publish to ghcr?

[the remark this answers is not in the document]
Comment (HonkingGoose) [reply .gitpod.Dockerfile:1]: > Can we move this to a `.gitpod` folder? so we don't pollute repo root to much?

The `.gitpod.yml` file must be in the root of the repo:

> A workspace gets configured through a `.gitpod.yml` file, located at the root of your project, written in YAML syntax. [^docs]

[^docs]: https://www.gitpod.io/docs/config-gitpod-file

File: .gitpod.Dockerfile (hunk 2 above)
[the remark this answers is not in the document]
Comment (HonkingGoose) [reply .gitpod.Dockerfile:5]: > Is the base image based on ubuntu? So we can install out buildpack files and use those installers to install our required versions?

Upstream source for `FROM gitpod/workspace-full`: https://github.com/gitpod-io/workspace-images/blob/master/full/Dockerfile

Upstream source for the `base` image: https://github.com/gitpod-io/workspace-images/blob/master/base/Dockerfile

File: .gitpod.yml [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+  file: .gitpod.Dockerfile
```
Comment (rarkins) [inline .gitpod.yml:2]: Could this file be moved under a subfolder, even if `.gitpod.yml` is at the root?

Comment (HonkingGoose) [reply .gitpod.yml:2]: I think the `gitpod.Dockerfile` must be in the root, at least that's what the docs seem to say: [^docs]

> ## Configure a custom Dockerfile
> 
> This option provides you with the most flexibility. Start by adding the following configuration in your .gitpod.yml file:
> 
> ```yaml
> image:
>   file: .gitpod.Dockerfile
> ```
> 
> Next, create a `.gitpod.Dockerfile` file at the root of your project. The syntax is the regular `Dockerfile` syntax as documented on docs.docker.com.

I have not tried if you can put the `.gitpod.Dockerfile` somewhere else and mention that in the `gitpod.yml` config...

[^docs]: https://www.gitpod.io/docs/config-docker#configure-a-custom-dockerfile

**Comment by HonkingGoose** [comment]
Maybe we're better off waiting until Gitpod supports `.devcontainer` files? Then we wouldn't need any extra files, or at least much less configuration work.

https://github.com/gitpod-io/gitpod/issues/7721

File: .gitpod.yml (hunk 5 above)
Comment (rarkins) [reply .gitpod.yml:2]: This comment implies it's possible: https://github.com/gitpod-io/gitpod/issues/7721#issuecomment-1018685624 (and maybe a different approach than this PR)

File: .gitpod.Dockerfile (hunk 2 above)
[the remark this answers is not in the document]
Comment (HonkingGoose) [reply .gitpod.Dockerfile:5]: Gitpod now comes with a version of Git that satisfies our current requirements, so we can probably drop the "install Git" stuff now?

**Comment by HonkingGoose** [comment]
It takes way, way too long to go from "building image" to "starting work". That's probably my fault, because I'm cobbling things together in the Dockerfile. 😄

When I start Gitpod from my fork, it gives me a default configuration, which works well for me, as:

- Husky checks for Prettier errors before committing
- I can run `yarn doc-fix` 
- I can run `brew install vale` to install the `vale` linter (#14835)
- I can run _some_ tests with `yarn jest`

I don't know enough Gitpod or Docker to get this PR in a state that we can merge it... 🙈 

How about we use the simple Gitpod configuration, and mention in a comment in the file that for "real development work" you should use the `.devcontainer` on GitHub Codespaces?

**Comment by HonkingGoose** [comment]
@rarkins How about just using a default Gitpod configuration instead of this complicated stuff? For must of the stuff I do, the default works just fine. We can add a comment to the gitpod file that for "real development work" you should use GitHub Codespaces, as that's what the maintainers use. 😄

File: .gitpod.Dockerfile (hunk 4 above)
[the remark this answers is not in the document]
Comment (viceice) [reply .gitpod.Dockerfile:1]: please try to move this to `.gitpod/Dockerfile` and report if not work

File: .gitpod.Dockerfile (hunk 2 above)
[the remark this answers is not in the document]
Comment (viceice) [reply .gitpod.Dockerfile:5]: Remove install stuff, but make sure we have node v14 or v16

is java preinstalled?

File: .gitpod.Dockerfile [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+
+# Install Node.js and use it
+
+RUN bash -c ". .nvm/nvm.sh && nvm install 14.18.1 && nvm use 14.18.1 && nvm alias default 14.18.1"
+
+RUN echo "nvm use default &>/dev/null" >> ~/.bashrc.d/51-nvm-fix
+
+# Install Python
+
+RUN bash -c "pyenv update && pyenv install 3.9.7 && pyenv global 3.9.7"
```
[the remark this answers is not in the document]
Comment (viceice) [reply .gitpod.Dockerfile:21]: We no longer need python as dev dependency

File: .gitpod.yml (hunk 5 above)
Comment (viceice) [reply .gitpod.yml:2]: Please try to move to subdir.
```suggestion
  file: .gitpod/Dockerfile
```

File: .gitpod.yml [hunk 7]
```diff
@@ -0,0 +1,7 @@
+# This configuration file was automatically generated by Gitpod.
```
Comment (HonkingGoose) [inline .gitpod.yml:1]: Moving files into a `.gitpod` directory causes Gitpod to just use it's default configuration file.

Let's simplify this whole thing, and start with the default Gitpod configuration file, and customize from there.

The tools that are currently installed in the default config are:

- `nvm --version`: `0.39.0`
- `node -v`: `16.13.2`
- `yarn --version`: `1.22.17`
- `git --version`: `2.35.1`
- `java --version`:
    ```
    Picked up JAVA_TOOL_OPTIONS:  -Xmx3435m
    openjdk 11.0.13 2021-10-19 LTS
    OpenJDK Runtime Environment Zulu11.52+13-CA (build 11.0.13+8-LTS)
    OpenJDK 64-Bit Server VM Zulu11.52+13-CA (build 11.0.13+8-LTS, mixed mode)
    ```
- `clang --version`:
  ```
  Ubuntu clang version 15.0.0-++20220226052900+e897214e97f2-1~exp1~20220226172958.171
  Target: x86_64-pc-linux-gnu
  Thread model: posix
  InstalledDir: /usr/bin
  ```
- `gcc --version`:
  ```
  gcc (Ubuntu 9.3.0-17ubuntu1~20.04) 9.3.0
  ```
- `gcmake --version`: `cmake version 3.22.3`

Comment (viceice) [reply .gitpod.yml:1]: sounds good, no need to customize

**Comment by HonkingGoose** [comment]
I've tried running `yarn jest` in Gitpod on the latest `main` and it's no good. I've aborted the run after giving it a really long time to finish. Here are the stats after I aborted the run:

```
$ yarn jest
......snip.....
Test Suites: 8 failed, 36 passed, 44 of 444 total
Tests:       78 failed, 4075 passed, 4153 total
Snapshots:   475 passed, 475 total
Time:        2150 s
```

Example of failing test output:

```
  ● modules/datasource/go/releases-goproxy › getReleases › supports "off" keyword

    thrown: "Exceeded timeout of 5000 ms for a test.
    Use jest.setTimeout(newTimeout) to increase the timeout value, if this is a long-running test."

      479 |     });
      480 |
    > 481 |     it('supports "off" keyword', async () => {
          |     ^
      482 |       process.env.GOPROXY = ['https://foo.com', 'https://bar.com', 'off'].join(
      483 |         ','
      484 |       );

      at lib/modules/datasource/go/releases-goproxy.spec.ts:481:5
      at lib/modules/datasource/go/releases-goproxy.spec.ts:266:3
      at Object.<anonymous> (lib/modules/datasource/go/releases-goproxy.spec.ts:7:1)
```

Maybe the Gitpod environment lacks the power to complete the Jest tests within the timeout, or something else is slowing it down.

---

I still need a basic Gitpod config so I can use Vale in the repository later (#14835).

We could put a comment in the development docs and the Gitpod config file to warn contributors to only use Gitpod for limited work.

```
Only use Gitpod to:

- Run Prettier
- Run ESLint
- Run Vale linter

Do not use Gitpod to run the Jest tests.
```

**Comment by HonkingGoose** [comment]
@rarkins I can't get the tests to pass when using Gitpod. Are you happy for me to use the default Gitpod configuration, and add a comment like this?

```yaml
# Use Gitpod for docs editing work only as the tests do not pass (yet).
# For development work, use GitHub Codespaces.
```

**Comment by viceice** [comment]
@HonkingGoose Is there node v16 on that image? then you need to use `yarn jest:16`, otherwise jest hungs with oom

**Comment by HonkingGoose** [comment]
Gitpod uses Node 16 on their default image. :)

**Comment by HonkingGoose** [comment]
Thank you @viceice, running `yarn jest:16` gives way better results! 🥳 

Still two failed tests though:

```bash
=============================== Coverage summary ===============================
Statements   : 99.98% ( 27079/27083 )
Branches     : 98.03% ( 8994/9174 )
Functions    : 100% ( 3285/3285 )
Lines        : 99.98% ( 26643/26647 )
================================================================================
Jest: "global" coverage threshold for statements (100%) not met: 99.98%
Jest: "global" coverage threshold for lines (100%) not met: 99.98%

Summary of all failing tests
 FAIL  lib/util/git/index.spec.ts (288.582 s, 209 MB heap size)
  ● util/git/index › isBranchModified() › should return false when author matches

    expected true to be false or Boolean(false)

      245 |
      246 |     it('should return false when author matches', async () => {
    > 247 |       expect(await git.isBranchModified('renovate/future_branch')).toBeFalse();
          |                                                                    ^
      248 |       expect(await git.isBranchModified('renovate/future_branch')).toBeFalse();
      249 |     });
      250 |

      at Object.<anonymous> (lib/util/git/index.spec.ts:247:68)
          at runMicrotasks (<anonymous>)

  ● util/git/index › isBranchModified() › should return false when author is ignored

    expected true to be false or Boolean(false)

      253 |         gitIgnoredAuthors: ['custom@example.com'],
      254 |       });
    > 255 |       expect(await git.isBranchModified('renovate/custom_author')).toBeFalse();
          |                                                                    ^
      256 |     });
      257 |
      258 |     it('should return true when custom author is unknown', async () => {

      at Object.<anonymous> (lib/util/git/index.spec.ts:255:68)
          at runMicrotasks (<anonymous>)


Test Suites: 1 failed, 453 passed, 454 total
Tests:       2 failed, 8732 passed, 8734 total
Snapshots:   1144 passed, 1144 total
Time:        406.327 s
Ran all test suites.
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

File: .gitpod.yml [hunk 8]
```diff
@@ -0,0 +1,7 @@
...
+# Please adjust to your needs (see https://www.gitpod.io/docs/config-gitpod-file)
+# and commit this file to your remote git repository to share the goodness with others.
+
+tasks:
```
Comment (HonkingGoose) [inline .gitpod.yml:5]: Do you want me to use `nvm` to "pin" the major version of Node.js to 16 in the `.gitpod.yml` file:

```yaml
tasks:
  - init: >
      nvm install 16 &&
      nvm use 16 &&
      yarn install &&
      yarn run build
    command: yarn run start
```

`nvm` will download the latest stable version of Node.js 16 each time the Gitpod container starts up. This takes 10 seconds or so, so it's not increasing the container start-up time much. And now the `major` version is pinned, so we don't get surprised when Gitpod starts using Node 17 for example.

**Comment by rarkins** [comment]
Just the git problem remaining?

File: docs/development/remote-development.md [hunk 9]
```diff
@@ -32,27 +37,33 @@ You can use [Gitpod](https://gitpod.io/) for light development work like:
...
 
 The Renovate developers use [GitHub Codespaces](https://github.com/features/codespaces).
 The config files are in the `.devcontainer` folder in the repository.
 
+You can only use GitHub Codespaces if you're a member of the Renovate organization, and have permission to use Codespaces.
+
 ### Codespaces tips
 
-List tips for codespaces here.
+List the tips for codespaces here.
```
Comment (HonkingGoose) [inline docs/development/remote-development.md:65]: Do you have tips for using Codespaces?

File: docs/development/remote-development.md [hunk 10]
```diff
@@ -32,27 +37,33 @@ You can use [Gitpod](https://gitpod.io/) for light development work like:
...
 
 ### Known problems with codespaces
 
-List known problems with codespaces here.
+List the known problems with codespaces here.
```
Comment (HonkingGoose) [inline docs/development/remote-development.md:69]: Are there known problems with Codespaces?

File: docs/development/remote-development.md [hunk 11]
```diff
@@ -32,27 +37,33 @@ You can use [Gitpod](https://gitpod.io/) for light development work like:
...
-`yarn jest:16` has some failing tests.
+- `yarn jest:16` has some failing tests
+- You can't preview Markdown files
 
 ## GitHub Codespaces
 
 The Renovate developers use [GitHub Codespaces](https://github.com/features/codespaces).
 The config files are in the `.devcontainer` folder in the repository.
 
+You can only use GitHub Codespaces if you're a member of the Renovate organization, and have permission to use Codespaces.
```
Comment (HonkingGoose) [inline docs/development/remote-development.md:61]: Please check if this is correct.

**Review by HonkingGoose — COMMENTED** [review]
I don't have access to GitHub Codespaces, so I need some help from you guys on the Codespaces parts. 😉

**Comment by HonkingGoose** [comment]
> Just the git problem remaining?

Pretty much. At least now nearly all test pass, previously they all failed, due to me using the wrong version of the `yarn :jest` command. 🙈 😄 

Do you _need_ to set a GitHub Personal Access token for all tests to pass?

What would you do if you set up a new development environment from scratch? I'm probably missing some vital step and that's why those two tests fail.

File: docs/development/remote-development.md (hunk 11 above)
Comment (rarkins) [reply docs/development/remote-development.md:61]: I think the beta may in fact be per-account. i.e. I think myself as @rarkins has the beta access and not @renovatebot the org. Possibly beta access is limited to renovatebot/renovate too

File: docs/development/remote-development.md (hunk 9 above)
Comment (rarkins) [reply docs/development/remote-development.md:65]: Let's leave it out for now so we can merge this doc earlier rather than later

**Review by viceice — DISMISSED** [review]

**Pull request merged**